x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,271 @@
+use swaprouter::msg::{ExecuteMsg as SwapRouterExecute, Slippage, SwapResponse};
+
+use crate::checks::{validate_memo, validate_receiver};
+use crate::consts::{CALLBACK_KEY, FORWARD_REPLY_ID, PACKET_LIFETIME, SWAP_REPLY_ID};
+use crate::ibc::{MsgTransfer, MsgTransferResponse};
+use crate::msg::{CrosschainSwapResponse, Recovery};
+
+use crate::state::{
+    ForwardMsgReplyState, ForwardTo, IBCTransfer, Status, SwapMsgReplyState, CONFIG,
+    FORWARD_REPLY_STATES, INFLIGHT_PACKETS, RECOVERY_STATES, SWAP_REPLY_STATES,
+};
+use crate::ContractError;
+
+/// This is the main execute call of this contract.
+///
+/// It's objective is to trigger a swap between the supplied pairs
+///
+#[allow(clippy::too_many_arguments)]
+pub fn swap_and_forward(
+    deps: DepsMut,
+    block_time: Timestamp,
+    contract_addr: Addr,
+    input_coin: Coin,
+    output_denom: String,
+    slippage: Slippage,
+    receiver: Addr,
+    next_memo: Option<String>,
+    failed_delivery: Option<Recovery>,
+) -> Result<Response, ContractError> {
+    deps.api.debug(&format!("executing swap and forward"));
+    let config = CONFIG.load(deps.storage)?;
+
+    // Message to swap tokens in the underlying swaprouter contract
+    let swap_msg = SwapRouterExecute::Swap {
+        input_coin: input_coin.clone(),
+        output_denom,
+        slippage,
+    };
+    let msg = wasm_execute(config.swap_contract, &swap_msg, vec![input_coin])?;
+
+    // Check that the received is valid and retrieve its channel
+    let (valid_channel, valid_receiver) = validate_receiver(deps.as_ref(), receiver)?;
+    // If there is a memo, check that it is valid
+    if let Some(memo) = &next_memo {
+        validate_memo(memo)?;
+    }
+
+    // Check that there isn't anything stored in SWAP_REPLY_STATES. If there is,
+    // it means that the contract is already waiting for a reply and should not
+    // override the stored state. This should only happen if a contract we call
+    // calls back to this one. This is likely a malicious attempt modify the
+    // contract's state before it has replied.
+    if SWAP_REPLY_STATES.may_load(deps.storage)?.is_some() {
+        return Err(ContractError::ContractLocked {
+            msg: "Already waiting for a reply".to_string(),
+        });
+    }
+    // Store information about the original message to be used in the reply
+    SWAP_REPLY_STATES.save(
+        deps.storage,
+        &SwapMsgReplyState {
+            swap_msg,
+            block_time,
+            contract_addr,
+            forward_to: ForwardTo {
+                channel: valid_channel,
+                receiver: valid_receiver,
+                next_memo,
+                failed_delivery,
+            },
+        },
+    )?;
+
+    Ok(Response::new().add_submessage(SubMsg::reply_on_success(msg, SWAP_REPLY_ID)))
+}
+
+pub fn handle_swap_reply(deps: DepsMut, msg: Reply) -> Result<Response, ContractError> {
```

Comment:
(annoying nit) Can we break some of this logic into smaller functions

Reply:
e.g. building response in case of no callback tracking, which seems like mostly boilerplate